client/x/evmengine/keeper/keeper.go

```diff
@@ -157,15 +158,38 @@ func (k *Keeper) parseAndVerifyProposedPayload(ctx context.Context, msg *types.M
 		return engine.ExecutableData{}, errors.New("invalid authority")
 	}
 
-	// validate execution payload
-	if err := types.ValidateExecPayload(msg); err != nil {
-		return engine.ExecutableData{}, errors.Wrap(err, "validate execution payload")
+	// Parse the payload.
+	var (
+		payload engine.ExecutableData
+		err     error
+	)
+
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	isV140, err := netconf.IsV140(sdkCtx.ChainID(), sdkCtx.BlockHeight())
+	if err != nil {
+		return engine.ExecutableData{}, errors.Wrap(err, "failed to check if the v1.4.0 upgrade is activated or not")
 	}
 
-	// Parse the payload.
-	var payload engine.ExecutableData
-	if err := json.Unmarshal(msg.ExecutionPayload, &payload); err != nil {
-		return engine.ExecutableData{}, errors.Wrap(err, "unmarshal payload")
+	//nolint:nestif // required to handle multiple conditional branches depending on v1.4.0 upgrade activation
+	if isV140 {
+		if msg.ExecutionPayload != nil {
+			return engine.ExecutableData{}, errors.New("legacy json payload not allowed")
+		}
+		payload, err = types.PayloadFromProto(msg.ExecutionPayloadDeneb)
+		if err != nil {
+			return engine.ExecutableData{}, errors.Wrap(err, "unmarshal proto payload")
+		}
+	} else {
+		if msg.ExecutionPayloadDeneb != nil {
+			return engine.ExecutableData{}, errors.New("proto payload not allowed")
+		}
+		// validate execution payload
+		if err := types.ValidateExecPayload(msg); err != nil {
+			return engine.ExecutableData{}, errors.Wrap(err, "validate execution payload")
+		}
+		if err := json.Unmarshal(msg.ExecutionPayload, &payload); err != nil {
+			return engine.ExecutableData{}, errors.Wrap(err, "unmarshal payload")
+		}
```

Comment:
Are we using the versioning branch check to make this binary backward-compatible (for a single binary?) If v1.4.0 is a hardfork, could we remove the else part since we don't need to keep the old decoding method?